Java-Inheritance.md

```diff
@@ -0,0 +1,168 @@
+
+### What can be inherited ?
+
+* All `protected` and `public` fields and methods from parent
+
+### What cannot be inherited ?
+
+* `private` fields and methods
+* Constructors. Although, subclass constructor _has_ to call superclass constructor (More on that later!)
+* Multiple classes. Java supports only Single Inheritance, that is you can only inherit one class at a time.
+* Fields. Individual fields of a class cannot be overriden by the subclass.
+
+## Type Casting & Reference
+
+In Java, it is possible to reference a subclass as an _instance_ of its superclass. For example, `Car` class object can be referenced as a `Vehicle` class instance like this : 
+
+```java
+Vehicle car = new Car();
+```
+
+Although, the opposite is not possible :
+
+```java
+Car car = new Vehicle(); // ERROR
+```
+
+So, since you can reference a Java subclass as a superclass instance, you can cast an subclass object easily to a superclass instance. It may be possible to cast a superclass object into a subclass type, but _only if the object is really an instance of subclass_. So keep this in mind :
+
+```java
+Car car = new Car();
+Vehicle vehicle = car; // upcasting
+Car car2 = (Car)vechile; //downcasting
+
+Bike bike = new Bike(); // say Bike is also a subclass of Vehicle
+Vehicle v = bike; // upcasting, no problem here.
+Car car3 = (Car)bike; // Compilation Error : as bike is NOT a instance of Car
+```
+
+Now you know how to share code through parent - child relationship. But, what if, you do not like the implementation of a particular method in the child class and want to write a new one for it? What to do then?
+
+## Override it!
+
+Java lets you override or redefine the methods defined in the superclass. For example, your `Car` class has a different implementation of `start()` than the parent `Vehicle`, so you do this :
+
+```java
+public class Vehicle() {
+	public void start() {
+	  System.out.println("Vehicle start code");
+	}
+}
+
+public class Car extends Vehicle {
+	public void start() {
+	  System.out.println("Car start code");
+  }
+}
+
+Car car = new Car();
+car.start(); // "Car start code"
+```
+
+So, its pretty simple to override methods in the subclass. Although, there is a _catch_. Only that superclass method will be overriden which has the _exact same method signature_ as the subclass method. Thus, `public void start(String key)` would not override `public void start()`.
+
+**Notes** :
+
+* You cannot override private methods of the superclass. (Quite obvious, isn't it?)
+* What if the method of superclass which you are overriding in the subclass suddenly gets obliterated or methods changed? It would fail in runtime! So Java provides you a nifty annotation `@Override` which you can place over the subclass method, which will warn the compiler of those incidents!
```

Comment:
Highlight the fact that that annontations like that in Java are good coding practice, not a necessity. The compiler is still smart enough to figure out overriding on its own.

Emphasize that unlike other OOP languages, like Python or TypeScript, it doesn't necessarily modify the method or add extra functionality.